src/potato_cluster.erl

```diff
@@ -0,0 +1,233 @@
+
+	    <<"now">> ->
+		Time = erlang:system_time(second),
+		json_put(genesis_block_timestamp_sec, Time, JsonConf_0);
+
+	    _ ->
+		JsonConf_0
+	end,
+
+    
+    
+    VerifierArr = pop_verifier:make_verifier_array_from_json(JsonConf),
+
+    VerAddressList = lists:map(
+		       fun(VerData) ->
+			       {VerAddress, _} = VerData#verifier_public_info.network_data,
+			       VerAddress
+		       end,
+		       array:to_list(VerifierArr)),
+
+    VerAddressUniqueList = sets:to_list(sets:from_list(VerAddressList)),
+
+    ProcessData = lists:map(
+		    fun(Address) ->
+			    UdpServerId = {udp_server, Address},
+			    IdList = start_single_server_cluster(VerifierArr, JsonConf, Address, UdpServerId, OnEventFn),
+			    {UdpServerId, IdList}
+		    end, 
+		    VerAddressUniqueList),
+
+    {UdpServerIdList, VerifierIdList} = lists:unzip(ProcessData),
+
+    {UdpServerIdList, lists:flatten(VerifierIdList)}.
+    
+stop_cluster({UdpServerIdList, VerifierIdList}) ->
+    lists:foreach(
+      fun(UdpId) ->
+	      gen_server:stop({global, UdpId})
+      end,
+      UdpServerIdList),
+
+    lists:foreach(
+      fun(Id) ->
+	      gen_server:stop({global, Id})
+      end,
+      VerifierIdList),
+
+    ok.
+
+
+format_message(full, Source, Code, Data) ->
+    io_lib:format("~p ~p ~n ~p ~n ~n", [Source, Code, Data]);
+
+format_message(code, Source, Code, _Data) ->
+    io_lib:format("~p ~n ~p ~n ~n", [Source, Code]);    
+
+format_message(json, Source, Code, Data) ->
+    S0 = io_lib:format("~p ~p ~n", [Source, Code]),
+    S1 = binary_to_list(jsx:encode(Data, [{indent, 4}, space])),
+    S2 = io_lib:format("~n~n", []),
+    S0 ++ S1 ++ S2.
+    
+
+block_chain_output(Source, Code = new_block_created, Block0, BlockLogFile) ->
+
+    CD0 = maps:get(consensus_data, Block0),
+    CD1 = maps:put(signature, omitted, CD0),
+    CD2 = maps:put(verifier_pub_key, omitted, CD1),
+
+    Block = maps:put(consensus_data, CD2, Block0),
+
+    file:write_file(BlockLogFile, format_message(json, Source, Code, Block), [append]);
+
+block_chain_output(_, _, _, _) ->
+    ok.
+
+
+start_web_cluster(Args) ->
+    try
+	start_web_cluster_inner(Args)
+    catch
+	_:Err ->
+	    Stacktrace = erlang:get_stacktrace(),
+	    ?debugFmt("~p ~n ~p ~n", [Err, Stacktrace]),
+	    erlang:error("can't start")
+    end.
+
+ 
+start_web_cluster_inner([JsonFileName, LogModeStr]) ->
+
+    %% ?debugFmt("~p ~n ~p ~n", [JsonFileName, LogModeStr]),
+
+    LogMode = LogModeStr,
+
+    {ok, FileData} = file:read_file(JsonFileName),
+
+    JsonConf = jsx:decode(FileData, [return_maps]),
+
+    case json_find(web, JsonConf) of
+	{ok, <<"none">>} ->
+	    no_web;
+	{ok, JsonWebConf} ->
+
+	    WebPort = json_get(port, JsonWebConf),
+	    WebLogsDir = json_get_str(logs_dir, JsonWebConf),
+	    WebFileDir = json_get_str(file_dir, JsonWebConf),
+
+	    potato_cluster_web_server:start(WebPort, WebLogsDir, WebFileDir),
+	    ok;
+	error ->
+	    no_web
+    end,
+
+    OnEventFn = 
```

Comment:
does this function log cluster/verifier ID?

Reply:
it does

Reply:
i think